### PR TITLE
Spatial Tranformer Networks

### LINKED ISSUE #107: Spatial transformer network (STN) layer
Hi,
I don't know if the STNs could be considered relevant enough to include them in this repo; however, I currently have an implementation of the STN layer [in this gist](https://gist.github.com/oarriaga/8438a7276bdc2b4ff03986d465e243b3). My original repository with a test case on the cluttered MNIST dataset can be found [here](https://github.com/oarriaga/spatial_transformer_networks). It is currently working with the tf backend (one can run the mnist_cluttered_example.ipynb). However, I recently moved everything into K tensors (in the K_tensors branch) but the `meshgrid`  from theano is failing. Looking at the `CONTRIBUTING.md` it seems like you only accept PRs that work with both back-ends, is this the case? Also, if someone is interested in having a look at the failing `K_meshgrid` function in the theano backend and helping me fix it, it would be awesome.

Thank you.

### PR DESCRIPTION
**Pull request by markroxor**
This is the implementation of [spatial transformer networks by the good people at DeepMind](https://arxiv.org/abs/1506.02025).

Quoting their abstract -  
> the use of spatial transformers results in models which learn invariance to translation, scale, rotation and more generic warping, resulting in state-of-the-art performance on several benchmarks, and for a number of classes of transformations.

I wanted to use STN for my research oriented tasks it would be a easier for me if I could directly import this layer from keras_contrib so that I could use it across my machines.

I have not updated the docs or written tests as it seems like the build is already failing. I will write tests once the master build is passed. I have tested the code locally on distorted mnist dataset. Perfect results.

fix #107 

EDIT: [Code credits](https://github.com/oarriaga/STN.keras/blob/master/src/models/layers.py) goes to @oarriaga

### CODE REVIEW

**Comment by markroxor** [comment]
@the-moliver If possible review it ASAP, I have work dependent on this PR.

**Comment by ahundt** [comment]
Thanks for the PR! You'll need to add unit tests and detailed comments on reference material as is done in the rest of the code before this can be merged.

**Comment by markroxor** [comment]
I think it would be a nice idea  if I can get the [example notebook](https://gist.github.com/markroxor/420c4c919ade739e5c2d3aae4d4eb8e6) merged with this PR as well. 

Since the usage is not that straight-forward.

File: keras_contrib/layers/spatial_transformer_networks.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+        affine_transformation = self.locnet.call(X)
+        output = self._transform(affine_transformation, X, self.output_size)
+        return output
+
+    def _repeat(self, x, num_repeats):
+        x = K.expand_dims(x, axis=-1)
+        x = K.repeat_elements(x, num_repeats, axis=1)
+        return K.reshape(x, [-1])
+
+    def _interpolate(self, image, x, y, output_size):
```
Comment (ahundt) [inline keras_contrib/layers/spatial_transformer_networks.py]: what kind of interpolation is this doing? needs comments

File: keras_contrib/layers/spatial_transformer_networks.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
+# -*- coding: utf-8 -*-
+from .. import backend as K
+from keras.engine import Layer
+
+if K.backend() == 'tensorflow':
```
Comment (ahundt) [inline keras_contrib/layers/spatial_transformer_networks.py]: This code needs to go into the specific backend's python file. Also maybe try adding mesh grid to keras itself?

File: keras_contrib/layers/spatial_transformer_networks.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        x_coordinates, y_coordinates = K_meshgrid(-1., 1., width,
+                                                  -1., 1., height)
+        x_coordinates = K.reshape(x_coordinates, [-1])
+        y_coordinates = K.reshape(y_coordinates, [-1])
+        ones = K.ones_like(x_coordinates)
+        indices_grid = K.concatenate([x_coordinates, y_coordinates, ones], 0)
+        return indices_grid
+
+    def _transform(self, affine_transformation, input_shape, output_size):
+        batch_size = K.shape(input_shape)[0]
```
Comment (ahundt) [inline keras_contrib/layers/spatial_transformer_networks.py]: what kind of transform, at least a brief comment would be helpful

File: keras_contrib/layers/spatial_transformer_networks.py (hunk 2 above)
Comment (markroxor) [reply keras_contrib/layers/spatial_transformer_networks.py]: you mean keras-team/keras?

Comment (markroxor) [reply keras_contrib/layers/spatial_transformer_networks.py]: Fixed other issues, I will need a little more elaboration on this one.

**Comment by oarriaga** [comment]
Hey @markroxor this very similar to my [STN code](https://github.com/oarriaga/STN.keras/blob/master/src/models/layers.py) could you please clarify what is the relationship to it. I just rewrote it such that there is less discrepancy between the backends.

File: keras_contrib/layers/spatial_transformer_networks.py (hunk 2 above)
Comment (oarriaga) [reply keras_contrib/layers/spatial_transformer_networks.py]: In the [newest implementation](https://github.com/oarriaga/STN.keras/blob/master/src/models/layers.py) there is now no need to have K_matmul defined for each backend.

**Comment by markroxor** [comment]
it is similar, actually as stated above I wanted it to be merged so that I can use it in my research work. I saw that your issue was not making any progress so I did it myself. Feel free to close this PR and submit your own :).

PS - Just get it merged ASAP please. :)

**Comment by mvaldenegro** [comment]
@markroxor The problem is that if this PR is are based on @oarriaga's code, then you should give proper credit. You could collaborate with it on getting the PR approved, but not giving credit is a serious problem. You should give it the proper attention, as is it not acceptable to take credit for other people's work.

**Comment by markroxor** [comment]
I have given code credits in the description. No plagiarism intended.

**Comment by oarriaga** [comment]
"similar" would be an understatement. I would please suggest to don't delete the contributors of the code which are all included in the original code from which you extracted yours. Also, you could easily copy the layer to your python file in which you are performing your training in order to easily  perform your research without the need of having this PR accepted.
My only remarks is not to delete the names of previous contributors, please do take that into consideration.

**Comment by markroxor** [comment]
I assume this PR should be closed. I apologize for the offences caused. I admit the mistake on my part.

**Comment by ahundt** [comment]
@markroxor Please respect the terms of peoples' license, whatever it is. In this case it is MIT and could hardly be less restrictive, and here is one key part:

```
The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.
```

The above copyright notice includes his name... @markroxor you need to bring your PR into compliance with the license, and in general you should always provide credit if you are using/extending others' work. If you properly credit this and @oarriaga is ok with it, we might proceed, though this situation does leave a bad taste... @oarriaga what do you think?

**Comment by dusk039** [comment]
@oarriaga do you plan to pursue this PR with proper credit? Great implementation by the way.

**Comment by markroxor** [comment]
@ahundt yours is the first polite response. I am relatively new to the community and wasn't expecting such rudeness. My first response to the situation was - "Feel free to close this PR and submit your own :)." but the discussion doesn't seem to end and I think before more people comment here, reciting the same thing over and over again, and make the matter worse I should clear my doubts -

I didn't want to bring it up but with all respect @ahundt @mvaldenegro @oarriaga, there are multiple implementations of STN over the internet which are strikingly "similar" (euphemism) to @oarriaga 's [code](https://github.com/oarriaga/STN.keras/blob/master/src/models/layers.py) - 
1. [In tensorflow](https://github.com/kevinzakka/spatial-transformer-network/blob/master/stn/transformer.py) following [MIT license](https://github.com/kevinzakka/spatial-transformer-network/blob/master/LICENSE)   
2. This one is interesting by [official Tensorflow](https://github.com/tensorflow/models/blob/master/research/transformer/spatial_transformer.py) itself under [Apache License 2.0](https://github.com/tensorflow/models/blob/master/LICENSE).

@ahundt Coming to the question - can we "own" the code by copying someone else's code and making some very slight modifications? (in this case even the function names are strikingly "similar" (euphemism)).

@ahundt I did not take the code from his MIT licensed repository but from his [unlicensed gist](https://gist.github.com/oarriaga/8438a7276bdc2b4ff03986d465e243b3), as I said I wanted to get the PR merged ASAP IDGAF about code credits. Even when a couple of days have passed I see there is no sign of getting a PR merged.

@oarriaga 
> Also, you could easily copy the layer to your python file in which you are performing your training in order to easily perform your research without the need of having this PR accepted.

I don't expect anyone I am sharing the code with to go and edit the site-packages to modify it and then use the layer. It is plain simple inconvenient. As a responsible developer/researcher it is your task to get the proposed work merged. Issue #107 is overdue for more than an year. Devs like @dusk039 and others who have commented on your issue can't wait to see STN in keras-contrib.

I would like this thread to be a future reference (for myself) in case I plan to "use" someone's code.
Expecting a polite response this time from @oarriaga  and his colleague @mvaldenegro .

**Comment by oarriaga** [comment]
Hello,

1.  I believe I have always responded politely.
2. My code is not a "very slightly modified" verion of David Dao's
     2.1 He gives credit in his repo to the actual first [implementation](https://github.com/skaae/transformer_network/blob/master/transformerlayer.py) on Lasagne.
    2.2 That implementation was then used by [Eder Santana's repo](https://github.com/EderSantana/seya)
    2.3 I based my code on Eder Santana's and I give credit to both implementations, but you don't.

3. My criticism was based on the fact that you deleted in your PR all previous contributors.
    3.2 Generally, any piece of code without a LICENSE should not be used and constrained to a LICENCE without reaching the original author.
    3.1 Copying code and erasing contributors displays a lack of appreciation and respect to all of those who worked without the direct intention of any economical retribution and display their code openly for others to use.

4. You can insert this layer in your code-base without having it explicitly in keras-contrib site-packages.

5. "As a responsible developer/researcher it is your task to get the proposed work merged" No, I believe it's not my task. 

5. I will continue my original PR since I believe as well that your test is incorrect.

**Comment by markroxor** [comment]
@ahundt Please lock the conversations.

**Comment by oarriaga** [comment]
The original gist you are referring to contains as well the original contributors.

**Pull request closed without merging**